File: 2025/pascal/day-04.pas
Program day_04;

{$mode objfpc}{$H+}{$J-}{$R+}{$Q+}

Uses SysUtils;

const
   INPUT =   '../input/day-04-input.txt';


type
   TGrid =  Array of Array of Boolean;
   TCoord =  Record
      y, x:  Integer;
   end;
   TCoordArray =  Array of TCoord;


function GetInput :  TGrid;
var
   tfIn:  textFile;
   s:  string;
   y, x:  Integer;
   grid:  TGrid;
begin
   AssignFile(tfIn, INPUT);
   SetLength(grid, 0);
   y := 0;

   try
      reset(tfIn);
      while not eof(tfIn) do
         begin
            ReadLn(tfIn, s);
            SetLength(grid, y + 1);
            SetLength(grid[y], Length(s));
            for x := 0 to Pred(Length(s)) do
               grid[y][x] := s[x + 1] = '@';
            Inc(y);
         end;
      CloseFile(tfIn);

   except
      on E:  EInOutError do WriteLn('IOerror: ', E.Message);
   end;

   Result := grid;
end;


function HasRoll(Const grid : TGrid; Const y, x: Integer) :  Boolean;
begin
   result := (y >= Low(grid)) And (x >= Low(grid))
             And (y <= High(grid))
             And (x <= High(grid[y]))
             And grid[y][x];
end;


function CheckAdjacent(Const grid : TGrid; Const y, x : Integer) :  Boolean;
var
   count, yy, xx:  Integer;
begin
   if not grid[y][x] then Exit(False);
   result := True;
   count := 0;
   for yy := y - 1 to y + 1 do
      for xx := x - 1 to x + 1 do
         begin
            if (yy = y) and (xx = x) then continue;
            if HasRoll(grid, yy, xx) then
               begin
                  Inc(count);
                  if count > 3 then Exit(False);
               end;
         end;
end;


procedure Part_1;
var
   grid:  TGrid;
   y, x, acc:  Integer;
begin
   grid := GetInput;
   acc := 0;
   for y := 0 to High(grid) do
      for x := 0 to High(grid[y]) do
         if CheckAdjacent(grid, y, x) then Inc(acc);
   WriteLn(acc);
end;


procedure Part_2;
var
   grid:  TGrid;
   mask:  TCoordArray;
   coord:  TCoord;
   y, x, cnt, acc:  Integer;
begin
   grid := GetInput;
   acc := 0;

   Repeat
      cnt := 0;
      SetLength(mask, Length(grid) * Length(grid[0]));

      for y := 0 to High(grid) do
         for x := 0 to High(grid[y]) do
            if CheckAdjacent(grid, y, x) then
               begin
                  coord.y := y;
                  coord.x := x;
                  mask[cnt] := coord;
                  Inc(cnt);
               end;

      if cnt = 0 then Break;
      Inc(acc, cnt);

      for coord in mask do
         grid[coord.y][coord.x] := false;

   Until false;

   WriteLn(acc);
end;

begin
   Part_1;
   Part_2;
end.

File: 2025/perl/day-04.pl
use v5.38;
use Path::Tiny;
use List::Util qw(first max);
use Data::Dump;

my $input = '../input/day-04-input.txt';

sub get_input {
    map [ map tr/@./10/r, split // ], path($input)->lines( { chomp => 1 } );
}

sub has_roll ( $grid, $y, $x ) {
    0 <= $y <= $#$grid
      and 0 <= $x <= $#{ $grid->[$y] }
      and $grid->[$y][$x];
}

sub check_adjacent ( $grid, $y, $x ) {
    return 0 if not $grid->[$y][$x];
    my $sount;
    for my $yy ( $y - 1 .. $y + 1 ) {
        for my $xx ( $x - 1 .. $x + 1 ) {
            next if $yy == $y && $xx == $x;
            return 0 if has_roll($grid, $yy, $xx) and ++$sount > 3;
        }
    }
    1;
}

sub part_1 {
    my @grid = get_input;
    my $count;
    for my $y ( keys @grid ) {
        for my $x ( keys $grid[$y]->@* ) {
            ++$count if check_adjacent \@grid, $y, $x;
        }
    }
    say $count;
}

sub part_2 {
    my @grid = get_input;
    my $count;

    while (1) {
        my @mask;

        for my $y ( keys @grid ) {
            for my $x ( keys $grid[$y]->@* ) {
                check_adjacent(\@grid, $y, $x) and push @mask, [ $y, $x ];
            }
        }

        last unless @mask;
        $count += @mask;
        undef $grid[ @$_[0] ][ @$_[1] ] for @mask;
    }
    say $count;
}

# dd get_input;

part_1;
part_2;
